Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp

```diff
@@ -152,51 +162,162 @@ ActsExamples::ProcessCode ActsExamples::TrackFindingAlgorithm::execute(
+                           trackCandidate.nTrackStates() - 1);
+            assert(firstState.previous() == Acts::kTrackIndexInvalid);
+
+            auto& secondTracksForSeed = secondResult.value();
+            for (auto& secondTrack : secondTracksForSeed) {
+              if (secondTrack.nTrackStates() < 2) {
+                continue;
+              }
+
+              // TODO a copy of the track should not be necessary but is the
+              //      safest way with the current EDM
+              // TODO a lightweight copy without copying all the track state
+              //      components might be a solution
+              auto secondTrackCopy = tracksTemp.makeTrack();
+              secondTrackCopy.copyFrom(secondTrack, true);
+
+              // Note that this is only valid if there are no branches
+              // We disallow this by breaking this look after a second track was
+              // processed
+              secondTrackCopy.reverseTrackStates(true);
+
+              firstState.previous() =
+                  (*std::next(secondTrackCopy.trackStatesReversed().begin()))
+                      .index();
+
+              Acts::calculateTrackQuantities(trackCandidate);
+
+              // TODO This extrapolation should not be necessary
+              // TODO The CKF is targeting this surface and should communicate
+              //      the resulting parameters
+              // TODO Removing this requires changes in the core CKF
+              //      implementation
+              auto secondExtrapolationResult =
+                  Acts::extrapolateTrackToReferenceSurface(
+                      trackCandidate, *pSurface, extrapolator,
+                      extrapolationOptions, m_cfg.extrapolationStrategy,
+                      logger());
+              if (!secondExtrapolationResult.ok()) {
+                m_nFailedExtrapolation++;
+                ACTS_ERROR("Second extrapolation for seed "
+                           << iSeed << " and track " << secondTrack.index()
+                           << " failed with error "
+                           << secondExtrapolationResult.error());
+
+                continue;
+              }
+
+              if (!m_trackSelector.has_value() ||
+                  m_trackSelector->isValidTrack(trackCandidate)) {
+                auto destProxy = tracks.makeTrack();
+                destProxy.copyFrom(trackCandidate, true);
+              }
+
+              ++nSecond;
+            }
+
+            // restore `trackCandidate` to its original state in case we need it
+            // again
+            firstState.previous() = Acts::kTrackIndexInvalid;
+            Acts::calculateTrackQuantities(trackCandidate);
+          }
+        }
+      }
 
-      if (!m_trackSelector.has_value() ||
-          m_trackSelector->isValidTrack(track)) {
-        auto destProxy = tracks.makeTrack();
-        destProxy.copyFrom(track, true);  // make sure we copy track states!
+      if (nSecond == 0) {
```

Comment:
Could you add a comment, when we will do the one-way-extrapolation? eg reasons why two-way might fail